Mathlib/Data/Ordmap/Ordnode.lean

```diff
@@ -0,0 +1,1365 @@
+  decide (size t₁ ≤ size t₂) && isSubsetAux t₁ t₂
+#align ordnode.is_subset Ordnode.isSubset
+
+/-- O(m + n). Is every element of `t₁` not equivalent to any element of `t₂`?
+
+     disjoint {1, 3} {2, 4} = tt
+     disjoint {1, 2} {2, 4} = ff -/
+def disjoint : Ordnode α → Ordnode α → Bool
+  | nil, _ => true
+  | _, nil => true
+  | node _ l x r, t =>
+    let (lt, found, gt) := split3 x t
+    found.isNone && disjoint l lt && disjoint r gt
+#align ordnode.disjoint Ordnode.disjoint
+
+/-- O(m * log(|m ∪ n| + 1)), m ≤ n. The union of two sets, preferring members of
+  `t₁` over those of `t₂` when equivalent elements are encountered.
+
+    union {1, 2} {2, 3} = {1, 2, 3}
+    union {1, 3} {2} = {1, 2, 3}
+
+Using a preorder on `ℕ × ℕ` that only compares the first coordinate:
+
+    union {(1, 1)} {(0, 1), (1, 2)} = {(0, 1), (1, 1)} -/
+def union : Ordnode α → Ordnode α → Ordnode α
+  | t₁, nil => t₁
+  | nil, t₂ => t₂
+  | t₁@(node s₁ l₁ x₁ r₁), t₂@(node s₂ _ x₂ _) =>
+    if s₂ = 1 then insert' x₂ t₁
+    else
+      if s₁ = 1 then insert x₁ t₂
+      else
+        let (l₂', r₂') := split x₁ t₂
+        link (union l₁ l₂') x₁ (union r₁ r₂')
+#align ordnode.union Ordnode.union
+
+/-- O(m * log(|m ∪ n| + 1)), m ≤ n. Difference of two sets.
+
+    diff {1, 2} {2, 3} = {1}
+    diff {1, 2, 3} {2} = {1, 3} -/
+def diff : Ordnode α → Ordnode α → Ordnode α
+  | t₁, nil => t₁
+  | t₁, t₂@(node _ l₂ x r₂) =>
+    cond t₁.empty t₂ <|
+      let (l₁, r₁) := split x t₁
+      let l₁₂ := diff l₁ l₂
+      let r₁₂ := diff r₁ r₂
+      if size l₁₂ + size r₁₂ = size t₁ then t₁ else merge l₁₂ r₁₂
+#align ordnode.diff Ordnode.diff
+
+/-- O(m * log(|m ∪ n| + 1)), m ≤ n. Intersection of two sets, preferring members of
+`t₁` over those of `t₂` when equivalent elements are encountered.
+
+    inter {1, 2} {2, 3} = {2}
+    inter {1, 3} {2} = ∅ -/
+def inter : Ordnode α → Ordnode α → Ordnode α
+  | nil, _ => nil
+  | t₁@(node _ l₁ x r₁), t₂ =>
+    cond t₂.empty t₁ <|
+      let (l₂, y, r₂) := split3 x t₂
+      let l₁₂ := inter l₁ l₂
+      let r₁₂ := inter r₁ r₂
+      cond y.isSome (link l₁₂ x r₁₂) (merge l₁₂ r₁₂)
+#align ordnode.inter Ordnode.inter
+
+/-- O(n * log n). Build a set from a list, preferring elements that appear earlier in the list
+in the case of equivalent elements.
+
+    ofList [1, 2, 3] = {1, 2, 3}
+    ofList [2, 1, 1, 3] = {1, 2, 3}
+
+Using a preorder on `ℕ × ℕ` that only compares the first coordinate:
+
+    ofList [(1, 1), (0, 1), (1, 2)] = {(0, 1), (1, 1)} -/
+def ofList (l : List α) : Ordnode α :=
+  l.foldr insert nil
+#align ordnode.of_list Ordnode.ofList
+
+/-- O(n * log n). Adaptively chooses between the linear and log-linear algorithm depending
+  on whether the input list is already sorted.
+
+    ofList' [1, 2, 3] = {1, 2, 3}
+    ofList' [2, 1, 1, 3] = {1, 2, 3} -/
+def ofList' : List α → Ordnode α
+  | [] => nil
+  | x :: xs => if List.Chain (fun a b => ¬b ≤ a) x xs then ofAscList (x :: xs) else ofList (x :: xs)
```

Comment:
This is the only thing that fails now with 
```
IR check failed at 'Ordnode.ofList'._rarg', error: unknown declaration 'List.decidableChain'
```

Reply:
pushed a workaround